### PR TITLE
Strengthen semantics of operator's life cycle hook

### PR DESCRIPTION
**Pull request by liuyehcf**
1. Ensure that all the stages of operator, including `prepare`、`set_finishing`、`set_finished`、`close`, will be exactly called once in the whole life cycle. And there is one exception that if prepare failed, then the other stages will never called.
2. And based on this, `set_finishing`、`set_finished`、`close` needn't to be idempotent any more
3. Add some life cycle assert to unit test  for the semantics of the `exactly once`

[related-issue](https://github.com/StarRocks/starrocks/issues/1292)

### CODE REVIEW

File: be/src/exec/pipeline/pipeline_driver.h [hunk 1]
```diff
@@ -176,6 +191,10 @@ class PipelineDriver {
 private:
     // check whether fragment is cancelled. It is used before pull_chunk and push_chunk.
     bool _check_fragment_is_canceled(RuntimeState* runtime_state);
+    void _finishing_operator(OperatorPtr& op, RuntimeState* runtime_state);
+    void _finish_operator(OperatorPtr& op, RuntimeState* runtime_state);
```
Comment (satanson) [inline be/src/exec/pipeline/pipeline_driver.h:195]: suggest to rename _finish_operator/_finishing_operator to _set_operator_finished/_set_operator_finishing.

finshing and finished  are participles and can be used as adjectives.

**Comment by satanson** [comment]
create issue then link this pr with it

File: be/src/exec/pipeline/pipeline_driver.cpp [hunk 2]
```diff
@@ -24,6 +24,7 @@ Status PipelineDriver::prepare(RuntimeState* runtime_state) {
     source_operator()->add_morsel_queue(_morsel_queue);
     for (auto& op : _operators) {
         RETURN_IF_ERROR(op->prepare(runtime_state));
+        _operator_stages[op->get_id()] = OperatorStage::PREPARED;
```
[the remark this answers is not in the document]
Comment (liuyehcf) [reply be/src/exec/pipeline/pipeline_driver.cpp:27]: Operator needn't maintain the field, it should be stored outside the operator

File: be/src/exec/pipeline/pipeline_driver.h (hunk 1 above)
Comment (liuyehcf) [reply be/src/exec/pipeline/pipeline_driver.h:195]: ok

File: be/src/exec/pipeline/pipeline_driver.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (liuyehcf) [reply be/src/exec/pipeline/pipeline_driver.cpp:27]: State is often used in state machines, it usually has very complex conversion logic. Stage expresses a more macro thing, the transition between stages is very simple. So I think stage is better

[the remark this answers is not in the document]
Comment (kangkaisen) [reply be/src/exec/pipeline/pipeline_driver.cpp:27]: stage +1

File: be/src/exec/pipeline/pipeline_driver.h [hunk 3]
```diff
@@ -195,6 +214,8 @@ class PipelineDriver {
     const size_t _yield_max_chunks_moved;
     const int64_t _yield_max_time_spent;
 
+    std::unordered_map<int32_t, OperatorStage> _operator_stages;
```
Comment (kangkaisen) [inline be/src/exec/pipeline/pipeline_driver.h:217]: Use phmap

File: be/src/exec/pipeline/pipeline_driver.h [hunk 4]
```diff
@@ -99,6 +99,17 @@ class DriverAcct {
     int64_t accumulated_chunk_moved{0};
 };
 
+// OperatorExecState is used to guarantee that some hooks of operator
+// is called exactly one time during whole operator
+enum OperatorStage {
+    INIT = 0,
```
Comment (kangkaisen) [inline be/src/exec/pipeline/pipeline_driver.h:105]: Do we really need the `INIT` stage?

File: be/src/exec/pipeline/pipeline_driver.h (hunk 3 above)
Comment (liuyehcf) [reply be/src/exec/pipeline/pipeline_driver.h:217]: ok

File: be/src/exec/pipeline/pipeline_driver.h (hunk 4 above)
Comment (liuyehcf) [reply be/src/exec/pipeline/pipeline_driver.h:105]: Not currently used, just for a complete definition

**Review by ZiheLiu — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

**Review by satanson — APPROVED** [review]

**Pull request merged**